strawberryfields/engine.py

```diff
@@ -495,13 +496,16 @@ def run(self, program, *, args=None, compile_options=None, **kwargs):
             program, args=args, compile_options=compile_options, **eng_run_options
         )
 
-        if valid_tdm_program:
-            result._all_samples = reshape_samples(
-                result.all_samples, program.measured_modes, program.N, program.timebins
-            )
-            # transpose the samples so that they have shape `(shots, spatial modes, timebins)`
-            result._samples = np.array(list(result.all_samples.values())).transpose(1, 0, 2)
-            program.roll()
+        if isinstance(program, TDMProgram):
+            if isinstance(result.all_samples, dict) and len(result.all_samples) > 0:
+                result._all_samples = reshape_samples(
+                    result.all_samples, program.measured_modes, program.N, program.timebins
+                )
+                # transpose the samples so that they have shape `(shots, spatial modes, timebins)`
+                result._samples = np.array(list(result.all_samples.values())).transpose(1, 0, 2)
+            if rolled_input:
+                # if tdm program is input in a rolled state, roll it back again
```

Comment:
Why? :thinking: I'd think if it's rolled, we don't need to roll it again :thinking:

Reply:
This means if the input program was rolled it should be kept rolled after the engine has done it's work. It will always have to be unrolled to simulate it.

So, it's either already unrolled (`rolled_input=False`) and then it's just used in it's unrolled state and not rolled again. Or it hasn't yet been unrolled (`rolled_input=True`), and then it must be unrolled and then rolled back again when done.

Reply:
Oh I see! That makes sense! :) Maybe it's worth phrasing the comment such that it states that we've now processed the unrolled circuit and want to restore its rolled state (if the input was rolled)? To me, that piece was missing, though maybe because I was looking at the change which is specific to the step.